lib/safe-http-client.js

```diff
@@ -0,0 +1,134 @@
+import dns from 'node:dns/promises'
+
+import {fetch} from 'undici'
+import ipaddr from 'ipaddr.js'
+
+import {IMAGE_MIME_TYPES} from './constants.js'
+
+export class HttpError extends Error {
+  /**
+   * @param {number} statusCode
+   * @param {string} message
+   */
+  constructor(statusCode, message) {
+    super(message)
+    this.statusCode = statusCode
+    this.name = 'HttpError'
+
+    // Captures the stack trace, excluding the constructor call from the stack trace.
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, HttpError)
+    }
+  }
+}
+
+export class SafeHttpClient {
+  /** @param {number} [maxSize] */
+  constructor(maxSize) {
+    this.maxSize = maxSize
+  }
+
+  /**
+   * Check if the URL is a valid URL or IP, that the prototol is valid,
+   * and the host is a safe unicast address.
+   * @param {string} url
+   */
+  static async checkUrl(url) {
+    // Throws if the URL is invalid
+    const validUrl = new URL(url)
+    const {protocol, hostname} = validUrl
+
+    // Don't allow aother protocols like file:// URLs
+    if (!['http:', 'https:'].includes(protocol)) {
+      throw new Error('Bad protocol')
+    }
```

Comment:
```suggestion
    // Only allow known safe protocols.
    if (!allowedProtocols.has(protocol)) {
      throw new Error('Bad protocol')
    }
```

Comment, and move a `const allowedProtocols = new Set(['http:', 'https:'])` to the top scope perhaps? As it’s a sort of constant?

Reply:
I think I prefer it the way it is. It's a simple inline check, no need to expose it globally in the file or function scope.

Reply:
I would go with this:

```suggestion
    // Don't allow aother protocols like file:// URLs
    if (protocol !== 'http:' && protocol !== 'https:') {
      throw new Error('Bad protocol')
    }
```

I’m not sure if v8 optimizes this, but it feels redundant to create an intermediate array.

The same pattern is repeated below.